### PR TITLE
LOG-1195: Add Openshift Ingest Plugin for ES

### PR DESCRIPTION
**Pull request by lukas-vlcek**
### Description
Configure Elasticsearch to install [Openshift Ingest Plugin](https://github.com/ViaQ/elasticsearch-openshift-ingest-plugin).

/cc @ewolinetz 
/assign @jcantrill 

### Links

- JIRA: <https://issues.redhat.com/browse/LOG-1195>

### CODE REVIEW

**Review by jcantrill — COMMENTED** [review]
/hold

File: elasticsearch/fetch-artifacts-koji.yaml [hunk 1]
```diff
@@ -1,3 +1,4 @@
 - nvr: org.elasticsearch-elasticsearch-6.8.1.redhat_00006-1
 - nvr: com.amazon.opendistroforelasticsearch-opendistro_security-0.10.0.4_redhat_00001-1
 - nvr: org.elasticsearch.plugin.prometheus-prometheus-exporter-6.8.1.0_redhat_00001-1
+- nvr: org.elasticsearch.plugin.ingest.openshift-ingest-plugin-6.8.1.0_redhat_00001-1
```
Comment (jcantrill) [inline elasticsearch/fetch-artifacts-koji.yaml:4]: This will break cpaas because it wont be found

Comment (lukas-vlcek) [reply elasticsearch/fetch-artifacts-koji.yaml:4]: I will comment it out for now... but once the PNC build it setup and running I can put it back, right?

**Comment by lukas-vlcek** [comment]
@jcantrill @ewolinetz I added default pipeline definition and initialization on ES startup. Also I defined the default pipeline in appropriate index templates. Please have a look again.

File: elasticsearch/ci-env.sh [hunk 2]
```diff
@@ -4,14 +4,19 @@ set -o xtrace
 
 PROMETHEUS_EXPORTER_URL=${PROMETHEUS_EXPORTER_URL:-$MAVEN_REPO_URL/org/elasticsearch/plugin/prometheus/prometheus-exporter/$PROMETHEUS_EXPORTER_VER/prometheus-exporter-$PROMETHEUS_EXPORTER_VER.zip}
 OPENDISTRO_URL=${OPENDISTRO_URL:-$MAVEN_REPO_URL/com/amazon/opendistroforelasticsearch/opendistro_security/$OPENDISTRO_VER/opendistro_security-$OPENDISTRO_VER.zip}
+INGEST_PLUGIN_URL=${INGESTI_PLUGIN_URL:-$MAVEN_REPO_URL/org/elasticsearch/plugin/ingest/openshift/$INGEST_PLUGIN_VER/openshift_ingest-plugin-$INGEST_PLUGIN_VER.zip}
```
Comment (ewolinetz) [inline elasticsearch/ci-env.sh:7]: s/INGESTI_PLUGIN_URL/INGEST_PLUGIN_URL/

File: elasticsearch/init/0100-seed-pipelines [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+source "logging"
+
+info Adding ingest pipelines
+shopt -s failglob
+for ingest_file in ${ES_HOME}/ingest_pipelines/*.json
+do
+    if [ -f ${ingest_file}.bkup ]; then
+      rm ${ingest_file}
+      mv ${ingest_file}.bkup ${ingest_file}
+    fi
```
Comment (ewolinetz) [inline elasticsearch/init/0100-seed-pipelines:27]: we shouldn't need this do we? we aren't doing any `sed` operations...

File: elasticsearch/fetch-artifacts-koji.yaml [hunk 4]
```diff
@@ -1,3 +1,4 @@
 - nvr: org.elasticsearch-elasticsearch-6.8.1.redhat_00006-1
 - nvr: com.amazon.opendistroforelasticsearch-opendistro_security-0.10.0.4_redhat_00001-1
 - nvr: org.elasticsearch.plugin.prometheus-prometheus-exporter-6.8.1.0_redhat_00001-1
+#- nvr: org.elasticsearch.plugin.ingest.openshift-ingest-plugin-6.8.1.0_redhat_00001-1
```
Comment (ewolinetz) [inline elasticsearch/fetch-artifacts-koji.yaml:4]: this is commented out because we don't yet have this being built in PNC, correct?

**Comment by ewolinetz** [comment]
> Q: Please let me know if I need to add the new folder `ingest_pipelines` on this line here:

we shouldn't need to because we aren't updating the file at all. no need to do writes there.

File: elasticsearch/init/0100-seed-pipelines (hunk 3 above)
Comment (lukas-vlcek) [reply elasticsearch/init/0100-seed-pipelines:27]: We do not. I will remove it and we can add later if needed.

File: elasticsearch/fetch-artifacts-koji.yaml (hunk 4 above)
Comment (lukas-vlcek) [reply elasticsearch/fetch-artifacts-koji.yaml:4]: corrent

File: elasticsearch/ci-env.sh (hunk 2 above)
Comment (lukas-vlcek) [reply elasticsearch/ci-env.sh]: fixed

File: elasticsearch/Dockerfile [hunk 5]
```diff
@@ -31,6 +31,7 @@ USER 0
 
 ARG ES_ARCHIVE_URL
 ARG PROMETHEUS_EXPORTER_URL
+ARG INGEST_PLUGIN_URL
```
Comment (jcantrill) [inline elasticsearch/Dockerfile:34]: add ```- nvr: org.elasticsearch.plugin.ingest-openshift-ingest-plugin-6.8.1.0_redhat_00001-1``` to ```https://github.com/openshift/origin-aggregated-logging/blob/master/elasticsearch/fetch-artifacts-koji.yaml```

File: elasticsearch/fetch-artifacts-koji.yaml (hunk 4 above)
Comment (jcantrill) [inline elasticsearch/fetch-artifacts-koji.yaml:4]: published version ```- nvr: org.elasticsearch.plugin.ingest-openshift-ingest-plugin-6.8.1.0_redhat_00001-1 ```

File: elasticsearch/ci-env.sh [hunk 6]
```diff
@@ -4,14 +4,19 @@ set -o xtrace
 
 PROMETHEUS_EXPORTER_URL=${PROMETHEUS_EXPORTER_URL:-$MAVEN_REPO_URL/org/elasticsearch/plugin/prometheus/prometheus-exporter/$PROMETHEUS_EXPORTER_VER/prometheus-exporter-$PROMETHEUS_EXPORTER_VER.zip}
 OPENDISTRO_URL=${OPENDISTRO_URL:-$MAVEN_REPO_URL/com/amazon/opendistroforelasticsearch/opendistro_security/$OPENDISTRO_VER/opendistro_security-$OPENDISTRO_VER.zip}
+INGEST_PLUGIN_URL=${INGEST_PLUGIN_URL:-$MAVEN_REPO_URL/org/elasticsearch/plugin/ingest/openshift/$INGEST_PLUGIN_VER/openshift_ingest-plugin-$INGEST_PLUGIN_VER.zip}
```
Comment (jcantrill) [inline elasticsearch/ci-env.sh:7]: If we never intend to pubblish this to an official maven repo then this line should simply default to  use the URL manipulation that you are doing for CI and/or pulling it straight from the Viaq repo

**Review by jcantrill — CHANGES_REQUESTED** [review]

**Comment by ewolinetz** [comment]
@lukas-vlcek can we also fix the `run.sh` file to properly pick up the log4j2.properties file?

File: elasticsearch/ci-env.sh (hunk 6 above)
Comment (lukas-vlcek) [reply elasticsearch/ci-env.sh:7]: This plugin is not generalized enough atm. It makes not sense to push it into any public mvn repo.

**Comment by lukas-vlcek** [comment]
@ewolinetz I think we can (though I would prefer separate PR).

File: elasticsearch/fetch-artifacts-koji.yaml [hunk 7]
```diff
@@ -1,3 +1,4 @@
 - nvr: org.elasticsearch-elasticsearch-6.8.1.redhat_00006-1
 - nvr: com.amazon.opendistroforelasticsearch-opendistro_security-0.10.0.4_redhat_00001-1
 - nvr: org.elasticsearch.plugin.prometheus-prometheus-exporter-6.8.1.0_redhat_00001-1
+#- nvr: org.elasticsearch.plugin.ingest-openshift-ingest-plugin-6.8.1.0_redhat_00001-1
```
Comment (lukas-vlcek) [inline elasticsearch/fetch-artifacts-koji.yaml:4]: @jcantrill When this line can be uncommented?

File: elasticsearch/Dockerfile.in [hunk 8]
```diff
@@ -57,6 +58,7 @@ ENV ES_PATH_CONF=/etc/elasticsearch/ \
     JAVA_HOME=/usr/lib/jvm/jre \
     NODE_QUORUM=1 \
     PROMETHEUS_EXPORTER_VER=6.8.1.0-redhat-00001 \
+    INGEST_PLUGIN_VER=6.8.1.0-redhat-00001 \
```
Comment (jcantrill) [inline elasticsearch/Dockerfile.in:61]: 6.8.1.0-redhat-00002

ref: https://brewweb.engineering.redhat.com/brew/buildinfo?buildID=1638435

File: elasticsearch/fetch-artifacts-koji.yaml [hunk 9]
```diff
@@ -1,3 +1,4 @@
 - nvr: org.elasticsearch-elasticsearch-6.8.1.redhat_00006-1
 - nvr: com.amazon.opendistroforelasticsearch-opendistro_security-0.10.0.4_redhat_00001-1
 - nvr: org.elasticsearch.plugin.prometheus-prometheus-exporter-6.8.1.0_redhat_00001-1
+- nvr: org.elasticsearch.plugin.ingest-openshift-ingest-plugin-6.8.1.0_redhat_00001-1
```
Comment (jcantrill) [inline elasticsearch/fetch-artifacts-koji.yaml:4]: org.elasticsearch.plugin.ingest-openshift-ingest-plugin-6.8.1.0_redhat_00002-1

https://brewweb.engineering.redhat.com/brew/buildinfo?buildID=1638435

File: elasticsearch/ci-env.sh [hunk 10]
```diff
@@ -4,14 +4,19 @@ set -o xtrace
...
 
 if [[ "${OPENSHIFT_CI:-}" == "true" ]]; then
     # This flag is set during CI runs. If no ARG was passed in,
     # default to maven.org.
     export ES_ARCHIVE_URL=https://github.com/openshift/origin-aggregated-logging/releases/download/elasticsearch-oss-$ES_VER/elasticsearch-oss-$ES_VER.zip
     export OPENDISTRO_URL=https://github.com/openshift/origin-aggregated-logging/releases/download/opendistro_security-$OPENDISTRO_VER/opendistro_security-$OPENDISTRO_VER.zip
 
+    # TODO[lvlcek]: We need to download official releases from ViaQ repo. Getting it from lukas-vlcek is temporary.
+    INGEST_PLUGIN_VER=$(echo $INGEST_PLUGIN_VER | cut -d'-' -f1)
+    export INGEST_PLUGIN_URL=https://github.com/lukas-vlcek/elasticsearch-openshift-ingest-plugin/releases/download/v$INGEST_PLUGIN_VER/openshift-ingest-plugin-$INGEST_PLUGIN_VER.zip
```
Comment (jcantrill) [inline elasticsearch/ci-env.sh:17]: Shouldnt we "correct" the upstream url to point to a release published here https://github.com/ViaQ/elasticsearch-openshift-ingest-plugin

File: elasticsearch/ci-env.sh (hunk 6 above)
Comment (jcantrill) [reply elasticsearch/ci-env.sh:7]: correct.  that is exactly my point.  we never intend to publish to maven so this change is incorrect (e.g. $MAVEN_REPO_URL)

File: elasticsearch/ci-env.sh (hunk 10 above)
Comment (lukas-vlcek) [reply elasticsearch/ci-env.sh:17]: Ok, I see what you mean @jcantrill
The thing is that we need to decide and control who, how and when cuts new releases for https://github.com/ViaQ/elasticsearch-openshift-ingest-plugin/

At this point it is just manual process. I created some tags in my clone for which I manually built releases and uploaded them to https://github.com/lukas-vlcek/elasticsearch-openshift-ingest-plugin/releases
(^^ This is what @vimalk78 used for his tests btw...)

Now, there is no process that is syncing tags between my clone and official ViaQ upstream.
We can either create tags manually or we can adopt similar approach that is used in prometheus-plugin – which means that whenever there is a commit that includes this property file without `-SNAPSHOT` suffix here ...
https://github.com/vvanholl/elasticsearch-prometheus-exporter/blob/master/gradle.properties#L3

... then a new tag is created automatically and a new release is built and uploaded automatically as well. See https://github.com/vvanholl/elasticsearch-prometheus-exporter/blob/master/.github/workflows/ci.yml#L39

We should decide how we want to control this. I would prefer fully automated approach (i.e. the samy way the prometheus plugin does it), but I am OK if you decide to do it differently.

//cc @ewolinetz

Comment (lukas-vlcek) [reply elasticsearch/ci-env.sh:17]: BTW, I do not think I have privs to create tags on ViaQ repo (the upstream). Which means if we decide we will create and upload release manually then I believe we need to ask @ewolinetz to do this.

**Pull request merged**